dao/dao.go

```diff
@@ -0,0 +1,320 @@
+	}
+	return nil
+}
+
+// Update the external of the SMT resource.
+func (dao *DAOImpl) UpdateSMTResourceExternalId(ctx context.Context, resourceId, externalId string) error {
+	_, err := GetClient().ReadWriteTransaction(ctx, func(ctx context.Context, txn *spanner.ReadWriteTransaction) error {
+		jobStmt := spanner.Statement{
+			SQL: `UPDATE SMT_RESOURCE SET ExternalId = @externalId, UpdatedAt = PENDING_COMMIT_TIMESTAMP()
+			WHERE ResourceId = @resourceId;`,
+			Params: map[string]interface{}{
+				"resourceId": resourceId,
+				"externalId": externalId,
+			},
+		}
+		_, err := txn.Update(ctx, jobStmt)
+		if err != nil {
+			return err
+		}
+		_, err = updateResourceHistoryWithinTxn(ctx, txn, resourceId)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("error updating smt resource external id: %v", err)
+	}
+	return nil
+}
+
+func updateJobHistoryWithinTxn(ctx context.Context, txn *spanner.ReadWriteTransaction, jobId string) (int64, error) {
+	version, err := getLatestJobVersionWithinTxn(ctx, txn, jobId)
+	if err != nil {
+		return 0, fmt.Errorf("error fetching latest job version: %v", err)
+	}
+	// Fetch the newly updated row from SMT_JOB table.
+	stmt := spanner.Statement{SQL: `
+		SELECT 
+			JobName, JobType, JobStateData, JobData, Dialect, SpannerDatabaseName 
+		FROM SMT_JOB WHERE JobId = @jobId;`,
+		Params: map[string]interface{}{"jobId": jobId},
+	}
+	iter := txn.Query(ctx, stmt)
+	defer iter.Stop()
+	var jobName, jobType, dialect, spannerDatabaseName spanner.NullString
+	var jobStateData, jobData spanner.NullJSON
+	row, err := iter.Next()
+	if err == iterator.Done || err != nil {
+		return 0, err
+	}
+	if err := row.Columns(&jobName, &jobType, &jobStateData, &jobData, &dialect, &spannerDatabaseName); err != nil {
+		return 0, fmt.Errorf("error reading smt job row: %v", err)
+	}
+
+	// Insert entry to SMT_JOB_HISTORY table.
+	jobStmt := spanner.Statement{
+		SQL: `INSERT INTO SMT_JOB_HISTORY 
+		(JobId, Version, JobName, JobType, JobStateData, JobData, Dialect, SpannerDatabaseName, CreatedAt)
+		VALUES(
+		 @jobId, @version, @jobName, @jobType, @jobStateData, @jobData, @dialect, @spannerDatabaseName, PENDING_COMMIT_TIMESTAMP()
+		);`,
+		Params: map[string]interface{}{
+			"jobId":               jobId,
+			"version":             version + 1,
+			"jobName":             jobName,
+			"jobType":             jobType,
+			"jobStateData":        jobStateData,
+			"jobData":             jobData,
+			"dialect":             dialect,
+			"spannerDatabaseName": spannerDatabaseName,
+		},
+	}
+	return txn.Update(ctx, jobStmt)
+}
+
+func getLatestJobVersionWithinTxn(ctx context.Context, txn *spanner.ReadWriteTransaction, jobId string) (int64, error) {
+	// Fetch latest version for the job from history table.
+	stmt := spanner.Statement{SQL: `SELECT MAX(Version) FROM SMT_JOB_HISTORY WHERE JobId = @jobId;`,
+		Params: map[string]interface{}{"jobId": jobId},
```

Comment:
This does not look like a good idea.

Monotonically increasing version numbers are a good idea when they can be generated via auto-increment by the DB itself. This kind of versioning (counting by 1) is an anti-pattern in Spanner, I think we should just make the `commit timestamp` as the other key in the PK instead of using a version number.

This will significantly simplify the DAO logic, avoid running `MAX` queries on the DB, will be human readable, and monotonic.


The same thing holds true for`resource history` as well.

Reply:
Even commit timestamp would cause hotspotting since that will also be monotonically increasing, but yes we can get rid of the fetch latest and increment logic in SMT.
Why did we decide to use an int version in the first place? I remember this was discussed i.e. having timestamp vs version col as PK.